### PR TITLE
implement a "fetch" cell for external resources (data, scripts, stylesheets)

### PR DESCRIPTION
**Pull request by bcolloran**
semi-working WIP fetch cell:

- can load JS files, but doesn't handle errors

still needs:
- load files (blob/text/json) to a variable
- load css
- handle errors of various sorts
- fetch local path files via editor

### CODE REVIEW

File: src/eval-frame/actions/fetch-cell-eval-actions.js [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+      .map(syntaxErrorToString)
+
+    if (syntaxErrors.length > 0) {
+      dispatch(appendToEvalHistory(
+        cell.id,
+        cell.content,
+        syntaxErrors,
+        { historyId, historyType: 'FETCH_CELL_INFO' },
+      ))
+      return Promise.resolve()
```
Comment (bcolloran) [inline src/eval-frame/actions/fetch-cell-eval-actions.js]: this should be a reject()

**Comment by hamilton** [comment]
I'm taking on this PR in the interest of time. At this point I believe everything should be covered:

- load javascript files as scripts (`js` fetch type)
- load and apply css files (`css` fetch type)
- load `text`/ `json` / `blob` types
- for all of the above, load either from an external endpoint (like a cdn) or from the iodide server (something like `files/<filename>`) through message passing

I'd like to do one last pass through this PR before requesting a review, since there is as lot of functionality implemented and I want to be sure I've covered all the edge cases.

### issues / room for improvement

1. Because we cannot incrementally print to the console (see #1107), I have to batch-update based on the current history item functionality. This means that even though everything is requested asynchronously, the update has to happen once everything has resolved one way or another. This seems very backward to me, and is a potential usability issue. If we fix #1107 we can fix this usability issue.
2. we should probably be keeping track of what libraries, files, etc. are loaded in the Workspace, more than just the declared variables.
3. this likely a trivial case for implementing our own CodeMirror syntax highlighting, which would drastically improve readability of these cell types.

**Comment by hamilton** [comment]
@wlach @bcolloran this is ready for a review. @bcolloran because this was your PR originally it doesn't look like I can add you as a reviewer.

**Comment by hamilton** [comment]
Here's some tests you might be able to run to check if it does everything. Some of these will fail unless you upload something to the server.

```
// text, json, blob: PASS CONDITION (remote)
text: test=https://cdnjs.cloudflare.com/ajax/libs/animate.css/3.7.0/animate.css
json: libraries= https://api.cdnjs.com/libraries
blob: gritty=https://media.giphy.com/media/MV1oM1koBZRvfp8Hvw/giphy.gif

// text, json, blob: FAIL CONDITION (remote)
text: animateFail=https://cdnjsasdsdfsdf.cloudflare.com/ajax/libs/animate.css/3.7.0/animate.css
blob: grittyFail=https://cdnsdfdsfjs.cloudflare.com/ajax/libs/animate.css/3.7.0/animate.cssssss
json: libFail=https://api.cdsdfnjs.com/librariessss

// text, json, blob: PASS CONDITION (if files uploaded to server)
text: testLocal= files/multiTimeline.csv // defines text: 
json: librariesLocal= files/libraries.json
blob: grittyLocal= files/gritty.mp4

// text, json, blob: FAIL CONDITION (if files uploaded to server)
text: testLocFail= files/multiTimeline.csvvvvv
json: libLocFail= files/libraries.jsonssss
blob: grittyLocFail= files/gritty.mp444444

// js: PASS CONDITIONS (server / remote)
js: https://cdnjs.cloudflare.com/ajax/libs/d3/5.7.0/d3.min.js
js: files/d3.min.js

// js: FAIL CONDITIONS (server / remote)
js: https://cccccccccdnjs.cloudflare.com/ajax/libs/d3/5.7.0/d3.min.js
js: filess/d3.min.js

// FAIL syntax errors in variable declaration
text: 9test=https://cdnjs.cloudflare.com/ajax/libs/animate.css/3.7.0/animate.css
json: 9libraries= https://api.cdnjs.com/libraries
blob: 9gritty=https://media.giphy.com/media/MV1oM1koBZRvfp8Hvw/giphy.gif

// FAIL syntax errors in fetch type should be caught.
1text: testLocal= files/multiTimeline.csv
jssson: librariesLocal= files/libraries.json
blobb: grittyLocal= files/gritty.mp4

// FAIL missing fetch type
testLocal= files/multiTimeline.csv
librariesLocal= files/libraries.json
grittyLocal= files/gritty.mp4

// css - PASS / FAIL conditions (server & remote)
css: https://gist.githubusercontent.com/hamilton/3afe20511547cb4f06ce5487a70702f1/raw/60df36c1f0c8109f5c73db6452012a08f69f7514/test.css
css: files/test-style.css

css: https://gist.111githubusercontent.com/hamilton/3afe20511547cb4f06ce5487a70702f1/raw/60df36c1f0c8109f5c73db6452012a08f69f7514/test.css
css: files11111/test-style.css
```

File: src/eval-frame/index.jsx [hunk 2]
```diff
@@ -25,6 +25,7 @@ import { iodide } from './iodide-api/api'
 import './port-to-editor'
 
 window.iodide = iodide
+window.FETCH_RESOLVERS = {}
```
Comment (wlach) [inline src/eval-frame/index.jsx]: Do we actually need to add a property to the window? Could we not define this object in a file and then import it elsewhere?

**Comment by wlach** [comment]
Rebased this branch on master for testing on heroku

File: src/eval-frame/index.jsx (hunk 2 above)
Comment (hamilton) [reply src/eval-frame/index.jsx]: Yep, it would be easy to have something that actually just closes over a variable in some way, along with some functions that fetches the `resolve` and `reject` as well as deletes.

Comment (hamilton) [reply src/eval-frame/index.jsx]: fwiw we use this pattern elsewhere, but agreed that maybe we should move to something that doesn't attach to `window`

**Comment by hamilton** [comment]
We're suspecting that a rebase may have caused some issue, so I'll investigate and fix any issues. Thanks for the review @wlach.

**Comment by wlach** [comment]
@hamilton seeing the same problem when deploying this to the heroku server as I did last friday, both console and report views seem to be empty. did you push all your changes?

**Comment by hamilton** [comment]
Hmm, I believe I did, but will check again in a few hours.

File: src/components/cells/cell-editor.jsx [hunk 3]
```diff
@@ -15,12 +15,23 @@ import 'codemirror/addon/display/autorefresh'
...
 import 'codemirror/addon/hint/show-hint'
 import 'codemirror/addon/hint/javascript-hint'
+import 'codemirror/addon/mode/simple'
 import '../../codemirror-keymap-sublime'
 
 import { getCellById } from '../../tools/notebook-utils'
 import * as actions from '../../actions/actions'
 import { postMessageToEvalFrame } from '../../port-to-eval-frame'
 
+CodeMirror.defineSimpleMode('fetch', {
```
Comment (bcolloran) [inline src/components/cells/cell-editor.jsx]: this is great! let's move it to a separate file tho

File: src/eval-frame/actions/__tests__/fetch-eval-actions.test.js [hunk 4]
```diff
@@ -0,0 +1,5 @@
+describe('stub', () => {
```
Comment (bcolloran) [inline src/eval-frame/actions/__tests__/fetch-eval-actions.test.js]: want to add some tests around "fetch-cell-eval-actions.js"? i put this stub in place, but for the stuff id had written so far i wasn't sure how to test it (fetch and promise stuff). if you know how to write tests around that, it would be awesome (it looks like some of these functions are pure enough that it might not be too bad?)

if not, let's delete this file.

File: src/components/cells/cell-editor.jsx (hunk 3 above)
Comment (hamilton) [reply src/components/cells/cell-editor.jsx]: yeah, was just thinking this earlier - will do.

File: src/eval-frame/actions/__tests__/fetch-eval-actions.test.js (hunk 4 above)
Comment (hamilton) [reply src/eval-frame/actions/__tests__/fetch-eval-actions.test.js]: Going to delete this file for now, I think. It's not immediately obvious how to test this one action as-is.

Comment (hamilton) [reply src/eval-frame/actions/__tests__/fetch-eval-actions.test.js]: That said, I do think it's worth getting the actions better tested (see #1054) I just don't know if this is the PR to begin tackling that.

File: src/eval-frame/actions/fetch-cell-parser.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+    // first, strip out comment from the end of line (if it exists)
+    const fetchCommand = fetchContent.trim().split(' //')[0].trim()
+    switch (fetchType) {
+      case 'text':
+        return Object.assign({}, { fetchType }, parseAssignmentCommand(fetchCommand))
+      case 'json':
+        return Object.assign({}, { fetchType }, parseAssignmentCommand(fetchCommand))
+      case 'blob':
+        return Object.assign({}, { fetchType }, parseAssignmentCommand(fetchCommand))
```
Comment (wlach) [inline src/eval-frame/actions/fetch-cell-parser.js]: I think you could shorten this to:

```js
case 'text':
case 'json':
case 'blob':
   return Object.assign({}, { fetchType }, parseAssignmentCommand(fetchCommand))
```

Likewise for the js/css cases below.

File: src/components/cells/cell-menu-container.js [hunk 6]
```diff
@@ -109,6 +109,8 @@ function mapStateToProps(state, ownProps) {
     label = 'resource'
   } else if (cell.cellType === 'raw') {
     label = 'raw'
+  } else if (cell.cellType === 'fetch') {
+    label = 'fetch'
```
Comment (wlach) [inline src/components/cells/cell-menu-container.js]: Could you simplify/generalize this code by using `cellTypeToJsmdMap` in jsmd-tools? (might be better done in a followup)

Comment (bcolloran) [reply src/components/cells/cell-menu-container.js]: nice catch will. i think it actually makes sense to do this now -- easier/cleaner/faster than having another issue+PR hanging over us

File: src/eval-frame/actions/fetch-cell-parser.js (hunk 5 above)
Comment (bcolloran) [reply src/eval-frame/actions/fetch-cell-parser.js]: +1 good catch

File: src/tools/fetch-tools.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  INVALID_VARIABLE_NAME: 'invalid variable name',
+}
+
+export function syntaxErrorToString(fetchInfo) {
+  return `Syntax error, ${errorTypeToString[fetchInfo.parsed.error]} in:
+      "${fetchInfo.line}"
+  `
+}
+
+export function handleErrors(err) {
```
Comment (wlach) [inline src/tools/fetch-tools.js]: Any reason this function needs to be exported?

File: src/tools/fetch-tools.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+export function successMessage(fetchInfo) {
+  const ifVarSet = fetchInfo.parsed.varName ? `\n\t(var ${fetchInfo.parsed.varName})` : ''
+  const text = `SUCCESS: ${fetchInfo.parsed.filePath} loaded${ifVarSet}\n\n`
+  return { text, id: fetchInfo.id }
+}
+
+export function errorMessage(fetchInfo, msg) {
+  let firstLine = ''
```
Comment (wlach) [inline src/tools/fetch-tools.js]: Probably would do:

```js
const firstLine = (fetchInfo.parsed.filePath) ? `${fetchInfo.parsed.filePath}\n\t` : '';
```

File: src/components/cells/cell-menu-container.js (hunk 6 above)
Comment (hamilton) [reply src/components/cells/cell-menu-container.js]: Sure, haven't really touched this so I'll have to look into what you mean :)

**Comment by bcolloran** [comment]
>     1. Because we cannot incrementally print to the console (see #1107), I have to batch-update based on the current history item functionality. This means that even though everything is requested asynchronously, the update has to happen once everything has resolved one way or another. This seems very backward to me, and is a potential usability issue. If we fix #1107 we can fix this usability issue.

we can update history items incrementally as events land, as we do in the plugin cell evaluation. that was my intent here, though i didn't get to that part. nonetheless, the update *does not* have to wate until everything is resolved. you can add this functionality with updateValueInHistory().

i think it's worth doing that now since you've identified this as a potential usability issue, and since this code is fresh in your mind, yes?

>     2. we should probably be keeping track of what libraries, files, etc. are loaded in the Workspace somehow.

can you elaborate? why do we need to track it?

**Comment by hamilton** [comment]
@bcolloran oh I see, you mean the syntax errors, not the failed fetches.

**Comment by bcolloran** [comment]
> @bcolloran you can't really have that & the async block fetching everything at the same time - there's no way to intercept an already-running fetch call. The best we can do is just surface that error. I figured it'd be better if a user knew all the errors present in the fetch cell, not just the first. Happy to change that if you feel otherwise.

@hamilton -- since the fetch cell parser does a pass over the complete cell contents, you can filter the parse results for syntax errors and surface them before kicking off any fetches

**Comment by hamilton** [comment]
@bcolloran right - so is the preferred behavior to show only the first error, or all the syntax errors?

**Comment by bcolloran** [comment]
i think we should surface all the syntax errors since we can catch them at parse time an halt evaluation before actually kicking off a bunch of potentially expensive fetches, and requiring the user to run a bunch of fetches twice. that was my thinking on it -- fail fast and early, ya know :-)

**Comment by hamilton** [comment]
@bcolloran sounds good, reread your original comment and saw "all" so I'm on it.

**Comment by hamilton** [comment]
@bcolloran addressed the issues you brought up. Let me know if there's anything else here.

**Comment by bcolloran** [comment]
woot!!! everything works as expected, solid code, great PR! thanks @hamilton, and nice work :-)
 :100:
